src/creature_groups.c

```diff
@@ -316,22 +316,91 @@ TbBool remove_creature_from_group_without_leader_consideration(struct Thing *cre
     return true;
 }
 
+/**
+ * Determines if a party has a Tunneler or Imp to consider for leadership.
+ * @param grptng is the creature whos party is considerd
+ * @return 0 if there's no digger, 1 if there's a digger who does not want to be a leader, and 2 if the digger is a preferred leader
+ */
+short creatures_group_has_special_digger_to_lead(struct Thing* grptng)
+{
+    struct Thing* ctng = INVALID_THING;
+    short potential_leader = 0;
+    struct CreatureControl* cctrl;
+    cctrl = creature_control_get_from_thing(grptng);
+    if (thing_is_creature_special_digger(grptng))
+    {
+        if (cctrl->party_objective != CHeroTsk_DefendParty)
+        {
+            potential_leader = 2;
+            return potential_leader;
+        }
+        else
+        {
+            potential_leader = 1;
+        }
+    }
+    long i = cctrl->group_info & TngGroup_LeaderIndex;
+    unsigned long k = 0;
+    if (i == 0)
+    {
+        i = grptng->index;
+    }
+    while (i > 0)
+    {
+        ctng = thing_get(i);
+        cctrl = creature_control_get_from_thing(ctng);
+        TRACE_THING(ctng);
+        if (thing_is_creature_special_digger(ctng))
+        {
+            if (cctrl->party_objective != CHeroTsk_DefendParty)
+            {
+                potential_leader = 2;
+                return potential_leader;
+            }
+            else
+            {
+                potential_leader = 1;
+            }
+        }
+        i = cctrl->next_in_group;
+        k++;
+        if (k > CREATURES_COUNT)
+        {
+            ERRORLOG("Infinite loop detected when sweeping creatures group");
+            break;
+        }
+    }
+    return potential_leader;
+}
+
+/**
+ * Finds a creature to become group leader. Considers objectives, diggers and score.
+ * @param grptng is the party member which needs a leader
+ * @return The Creature that should lead the group.
+ */
 struct Thing* get_best_creature_to_lead_group(struct Thing* grptng)
```

Comment:
If I did the whole thing I might've refactored the existing one to only do a loop over all party members to determine who has the highest "leader_score" and a separate function to determine said score like the most significant bit in said score would be if it's a tunneler, the next bit if it's not defending, then 8bits for the creature_thing_score and the last few bits for the explevel etc, would be bit cleaner, current function as is doesn't fit on my screen, which is an indication that it could be better